java/java-tests/testData/inspection/dataFlow/fixture/ConsumedStream.java

```diff
@@ -0,0 +1,191 @@
+  }
+
+  public static void test3E() {
+    Stream<String> stream = Stream.of("x");
+    double random = Math.random();
+
+    if (random > 0.5) {
+      stream.forEach(System.out::println);
+    }
+
+    if (random > 0.9) {
+      <warning descr="Stream has already been linked or consumed">stream</warning>.forEach(System.out::println);
+    }
+  }
+
+  public static void test6E() {
+    Stream<String> stream = Stream.of("x");
+    double random2 = Math.random();
+    Stream<String> stringStream = stream.filter(t -> true);
+    boolean parallel = stream.isParallel();
+    double random3 = Math.random();
+    method(stream);
+    <warning descr="Stream has already been linked or consumed">stream</warning>.filter(t -> true);
+  }
+
+  public static void test7E() {
+    Stream<String> stream = Stream.of("x");
+    Stream<String> parallel = stream.parallel().filter(t->true);
+    <warning descr="Stream has already been linked or consumed">stream</warning>.forEach(System.out::println);
+  }
+
+  public static void test8E() {
+    Stream<String> stream = Stream.of("x");
+    stream.forEach(System.out::println);
+    Stream<String> parallel = stream.parallel();
+    <warning descr="Stream has already been linked or consumed">parallel</warning>.forEach(System.out::println);
+  }
+
+  private static void method(Stream<String> stream) {
+    Iterator<String> iterator = stream.iterator();
+  }
+
+  public static void test1N() {
+    Stream<String> stream = Stream.of("x");
+    if (Math.random() > 0.5) {
+      stream.forEach(System.out::println);
+    } else {
+      stream.distinct().forEach(System.out::println);
+    }
+  }
+
+  public static void test2N() {
+    Stream<String> stream = Stream.of("x");
+    stream = stream.filter(x -> !x.isEmpty());
+    stream.forEach(System.out::println);
+  }
+
+  public static void test3N() {
+    Stream<String> stream = Stream.of("x");
+    System.out.println(stream.isParallel());
+    stream.forEach(System.out::println);
+  }
+
+  public static void test4N() {
+    Stream<String> stream = Stream.of("x");
+    switch (new Random().nextInt(10)) {
+      case 0:
+        stream.forEach(System.out::println);
+        break;
+      case 1:
+        stream.forEachOrdered(System.out::println);
+    }
+  }
+
+  public static void test5N() {
+    Stream<String> stream = Stream.of("x");
+    for (int i = 0; i < 10; i++) {
+      if (Math.random() > 0.5) {
+        stream.forEach(System.out::println);
+        break;
+      }
+    }
+  }
+
+  public static void test1NE() {
+    Stream<String> stream = Stream.of("x");
+    boolean b = Math.random() > 0.5;
+    if (b) {
+      stream.forEach(System.out::println);
+    }
+    if (!b) {
+      stream.forEach(System.out::println);
+    }
+  }
+
+  public static void test2NE() {
+    Stream<String> stream = Stream.of("x");
+    for (int i = 0; i < 10; i++) {
+      if (i == 5) {
+        stream.forEach(System.out::println);
```

Comment:
I wonder how this works... Do we really know that `i == 5` will happen only once in the whole loop? I don't remember that our analysis supports this. When if we change the condition to `if (i == 5 || i == 7)`?

Reply:
Loops are not supported exactly. The case `if (i == 5 || i == 7)` doesn't show warning too. It happens, because the interpreter tries to squash 2 state (1 - when `i == 5`, 2 - when `i != 5`) and `stream.linkedOrConsumed` becomes `BOOLEAN` (or `UNKNOWN` now).
This behavior could be changed, but I've decided to save it, because it can reduce false-positive cases. If it is required, I will fix it

Reply:
Well, `i == 5` test was suggested in case you decide to implement this as a separate inspection (matching PSI), though it would work worse in other cases. We may try to fix this later somehow in general, making dataflow analysis to know that in counting loops values don't repeat. This is definitely out of the scope of this test assignment.

Still probably it's good to report this problem (falsely) for now, as the fact of squashing is the implementation detail of the analysis and should not change the warnings. Well, it's up to you, for now you can leave it as is.